File: Sources/App/Classes/Dialogs/TDCReactionPopoverController.m
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\___/_/\_\\__|\__,_|\__,_|_|
 *
 * Copyright (c) 2010 - 2026 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of Textual, "Codeux Software, LLC", nor the
 *    names of its contributors may be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *********************************************************************** */

#import "TLOLocalization.h"
#import "TDCReactionPopoverControllerPrivate.h"

NS_ASSUME_NONNULL_BEGIN

@interface TDCReactionPopoverController () <NSPopoverDelegate, NSTextFieldDelegate>
@property(nonatomic, copy, readwrite) NSString *messageIdentifier;
@property(nonatomic, strong, nullable) NSPopover *popover;
@property(nonatomic, strong) NSTextField *emojiField;
@property(nonatomic, strong) NSButton *sendButton;
@end

@implementation TDCReactionPopoverController

- (instancetype)initWithMessageIdentifier:(NSString *)messageIdentifier
{
	NSParameterAssert(messageIdentifier != nil);

	if ((self = [super initWithNibName:nil bundle:nil])) {
		self.messageIdentifier = messageIdentifier;
	}

	return self;
}

- (void)loadView
{
	NSView *view = [[NSView alloc] initWithFrame:NSMakeRect(0.0, 0.0, 220.0, 44.0)];

	NSTextField *field = [[NSTextField alloc] initWithFrame:NSZeroRect];

	field.translatesAutoresizingMaskIntoConstraints = NO;
	field.placeholderString = TXTLS(@"TXMenuController[rct-ph]");
	field.font = [NSFont systemFontOfSize:16.0];
	field.alignment = NSTextAlignmentCenter;
	field.delegate = self;
	field.bezelStyle = NSTextFieldRoundedBezel;
	field.usesSingleLineMode = YES;

	NSButton *send = [NSButton buttonWithTitle:TXTLS(@"TXMenuController[rct-sd]") target:self action:@selector(send:)];

	send.translatesAutoresizingMaskIntoConstraints = NO;
	send.bezelStyle = NSBezelStyleRounded;
	send.keyEquivalent = @"\r";
	send.enabled = NO;

	[view addSubview:field];
	[view addSubview:send];

	[NSLayoutConstraint activateConstraints:@[
		[field.leadingAnchor constraintEqualToAnchor:view.leadingAnchor constant:12.0],
		[field.centerYAnchor constraintEqualToAnchor:view.centerYAnchor],
		[field.widthAnchor constraintEqualToConstant:120.0],

		[send.leadingAnchor constraintEqualToAnchor:field.trailingAnchor constant:8.0],
		[send.trailingAnchor constraintEqualToAnchor:view.trailingAnchor constant:-12.0],
		[send.firstBaselineAnchor constraintEqualToAnchor:field.firstBaselineAnchor],

		[view.heightAnchor constraintEqualToConstant:44.0],
	]];

	self.emojiField = field;
	self.sendButton = send;

	self.view = view;
}

- (void)presentRelativeToRect:(NSRect)rect ofView:(NSView *)view
{
	NSParameterAssert(view != nil);

	NSPopover *popover = [NSPopover new];

	popover.behavior = NSPopoverBehaviorTransient;
	popover.contentViewController = self;
	popover.delegate = self;

	self.popover = popover;

	[popover showRelativeToRect:rect ofView:view preferredEdge:NSRectEdgeMaxY];

	[self.view.window makeFirstResponder:self.emojiField];
}

- (void)close
{
	[self.popover close];
}

- (void)popoverDidClose:(NSNotification *)notification
{
	self.popover = nil;
}

/* One emoji: the first grapheme cluster of what was typed. */
- (nullable NSString *)emoji
{
	NSString *value =
		[self.emojiField.stringValue stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];

	if (value.length == 0) {
		return nil;
	}

	NSRange first = [value rangeOfComposedCharacterSequenceAtIndex:0];

	return [value substringWithRange:first];
}

- (void)controlTextDidChange:(NSNotification *)notification
{
	self.sendButton.enabled = (self.emoji != nil);
}

- (void)send:(nullable id)sender
{
	NSString *emoji = self.emoji;

	if (emoji == nil) {
		return;
	}

	if (self.completionBlock) {
		self.completionBlock(emoji, self.messageIdentifier);
	}

	[self close];
}

@end

NS_ASSUME_NONNULL_END

File: Sources/App/Modules/AppKitSupport/TDCReactionPopoverController.swift
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \\ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\\___/_/\\_\\__|\\__,_|\\__,_|_|
 *
 * Copyright (c) 2010 - 2026 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 *********************************************************************** */

import AppKit

@objc(TDCReactionPopoverController)
@MainActor
public final class ReactionPopoverController: NSViewController, NSPopoverDelegate, NSTextFieldDelegate {
	@objc public private(set) var messageIdentifier: String
	@objc public var completionBlock: ((String, String) -> Void)?

	private var popover: NSPopover?
	private var emojiField: NSTextField!
	private var sendButton: NSButton!

	@objc(initWithMessageIdentifier:)
	public init(messageIdentifier: String) {
		self.messageIdentifier = messageIdentifier
		super.init(nibName: nil, bundle: nil)
	}

	@available(*, unavailable)
	public required init?(coder _: NSCoder) {
		fatalError("init(coder:) has not been implemented")
	}

	override public func loadView() {
		let view = NSView(frame: NSRect(x: 0, y: 0, width: 220, height: 44))

		let field = NSTextField(frame: .zero)
		field.translatesAutoresizingMaskIntoConstraints = false
		field.placeholderString = LocalizedKey("TXMenuController[rct-ph]")
		field.font = NSFont.systemFont(ofSize: 16.0)
		field.alignment = .center
		field.delegate = self
		field.bezelStyle = .roundedBezel
		field.usesSingleLineMode = true

		let send = NSButton(
			title: LocalizedKey("TXMenuController[rct-sd]"),
			target: self,
			action: #selector(send(_:))
		)
		send.translatesAutoresizingMaskIntoConstraints = false
		send.bezelStyle = .rounded
		send.keyEquivalent = "\r"
		send.isEnabled = false

		view.addSubview(field)
		view.addSubview(send)

		NSLayoutConstraint.activate([
			field.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 12.0),
			field.centerYAnchor.constraint(equalTo: view.centerYAnchor),
			field.widthAnchor.constraint(equalToConstant: 120.0),

			send.leadingAnchor.constraint(equalTo: field.trailingAnchor, constant: 8.0),
			send.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -12.0),
			send.firstBaselineAnchor.constraint(equalTo: field.firstBaselineAnchor),

			view.heightAnchor.constraint(equalToConstant: 44.0),
		])

		emojiField = field
		sendButton = send
		self.view = view
	}

	@objc(presentRelativeToRect:ofView:)
	public func present(relativeTo rect: NSRect, of view: NSView) {
		let popover = NSPopover()
		popover.behavior = .transient
		popover.contentViewController = self
		popover.delegate = self

		self.popover = popover
		popover.show(relativeTo: rect, of: view, preferredEdge: .maxY)

		self.view.window?.makeFirstResponder(emojiField)
	}

	@objc public func close() {
		popover?.close()
	}

	public func popoverDidClose(_: Notification) {
		popover = nil
	}

	private var emoji: String? {
		let value = emojiField.stringValue.trimmingCharacters(in: .whitespacesAndNewlines)

		guard value.isEmpty == false else {
			return nil
		}

		let first = value.rangeOfComposedCharacterSequence(at: value.startIndex)
		return String(value[first])
	}

	public func controlTextDidChange(_: Notification) {
		sendButton.isEnabled = (emoji != nil)
	}

	@IBAction public func send(_: Any?) {
		guard let emoji else {
			return
		}

		completionBlock?(emoji, messageIdentifier)
		close()
	}
}
